### PR TITLE
feat(chat): remember JSON output view mode

### LINKED ISSUE #3071: [Feature Request] Remember command JSON view mode across sessions
### What should we add/change?

When a command returns JSON, selecting Summary, Tree, or Raw should update the preferred JSON view mode. New command outputs should use that preference across every session instead of resetting to Summary.

Keep Summary as the default when no preference has been saved.

### Extra context

The expandable tool output currently keeps `jsonViewMode` as component-local state and resets it to `summary` whenever output changes. Persisting the selected mode in the global UI preferences would remove repeated toggling without changing non-JSON tool output.

### PR DESCRIPTION
**Pull request by karimodm**
## Intent

Closes #3071.

Remember the Summary, Tree, or Raw mode selected for expandable command JSON output. The preference lives in `useUIStore` and follows the existing settings persistence path. New, refreshed, and already-mounted JSON cards read the same value.

The web server sanitizer accepts the three supported values, so settings saves and reloads round-trip the preference instead of resetting it to Summary.

## Non-goals

This PR does not add a Settings control, change the separate file or full-screen JSON viewers, or add a new synchronization service.

## Affected surfaces

`packages/ui` owns the preference, tool-card subscription, default, hydration, and autosave wiring. Web, desktop, VS Code, hosted mobile, and Capacitor mobile use that shared store behavior.

`packages/web` owns the settings sanitizer used by settings writes and reads. The sanitizer accepts only `summary`, `formatted`, or `raw`. Existing settings without the field continue to use Summary. The bespoke localStorage key and `window.top` adapter are not part of the final design.

## Repository guidance

| Guidance | Why it applies | How the change complies |
|---|---|---|
| `AGENTS.md`, `CONTRIBUTING.md`, and `openchamber-change-discipline` | This changes shared executable UI and a persisted settings contract. | The preference uses the existing store and settings round-trip. Focused client and server tests cover defaults, valid values, invalid values, and autosave. |
| `sync-state-invariants`, `packages/ui/src/sync/DOCUMENTATION.md`, and `packages/ui/src/stores/DOCUMENTATION.md` | One store must own the preference across mounted chat cards and settings hydration. | `useUIStore` owns the value. Existing persistence and settings application code handle rehydration, authoritative omission, and autosave. |
| `ui-api-decoupling`, its implementation map, and `packages/web/server/lib/opencode/DOCUMENTATION.md` | The preference crosses the shared UI and server settings boundary. | The existing settings route remains the transport boundary. Client and server accept the same three-value contract. |
| `packages/ui/src/components/chat/message/parts/DOCUMENTATION.md` | This document owns expandable JSON tool output behavior. | It points to persisted UI settings and documents the Summary fallback. |
| `communication-style` | The PR description and component documentation are maintainer-facing text. | The handoff describes the current store-based design and its limits directly. |

## Validation

Integrated HEAD: `4a13951b9d6a3be04a826b4ad5f600c73359c015`. The original change was merged with main `c8e0ea68f` without conflicts. The final feature delta remains 10 files, +168/-85.

| Check | Result |
|---|---|
| `bun test packages/web/server/lib/opencode/settings-helpers.test.js` | 40 passed. |
| `bun test packages/ui/src/lib/persistence.test.ts` | 37 passed. |
| `bun test packages/ui/src/components/chat/message/parts/ToolPart.test.ts` | 31 passed. |
| `bun test packages/ui/src/lib/i18n/messages.test.ts` | 2 passed. The previous missing Turkish-key failure is resolved by current main. |
| `bun run type-check` | Passed in every workspace. |
| `bun run lint` | Passed, with one pre-existing React hook warning in MobileChangesSurface.tsx. |
| `bun run build` | Passed, including web, VS Code, and mobile assets. Existing bundle-size/font warnings remain. |
| `bun run dead-code` | Completed and inspected. Existing unused-code report; no new JSON-view symbol reported. |
| `bunx --no-install oxlint <changed JS/TS files>` | Nonzero on existing anti-slop backlog. Two diagnostics overlap moved, unchanged ToolScrollableTextOutput declarations; none concern newly authored JSON-preference statements. |
| `node --check packages/web/server/lib/opencode/settings-helpers.js` | Passed. |
| `git diff --check` | Passed. |
| Full test suite | Passed in the required [GitHub checks run](https://github.com/openchamber/openchamber/actions/runs/33974480536) at this HEAD, along with build, type-check, lint, generated-changelog verification, and Electron Linux packaging tests. |
| Browser interaction | Not run locally. |

## Visual evidence

There is no styling or layout change. The existing Summary, Tree, and Raw controls render as before. The changed behavior is that selecting one updates other mounted JSON cards and survives a settings reload.

## Risks and failure behavior

Missing or invalid settings values use Summary. The client and server accept the same three values. The server sanitizer prevents malformed values from entering the settings file, and client hydration keeps the default when an older settings file omits the field.

Every mounted JSON card subscribes to one small store field. Selecting a mode updates those cards immediately and schedules persistence through the existing appearance autosave path. Reverting the commits removes the field; older settings remain readable because unknown values are sanitized by the existing settings code.

### CODE REVIEW

File: packages/ui/src/components/chat/message/parts/toolJsonViewPreference.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    return 'summary';
+};
+
+export const getToolJsonViewMode = (): ToolJsonViewMode => (
+    parseToolJsonViewMode(getSafeStorage().getItem(TOOL_JSON_VIEW_MODE_KEY))
+);
+
+export const setToolJsonViewMode = (value: ToolJsonViewMode): void => {
+    getSafeStorage().setItem(TOOL_JSON_VIEW_MODE_KEY, value);
+};
```
[the remark this answers is not in the document]
Comment (karimodm) [reply packages/ui/src/components/chat/message/parts/toolJsonViewPreference.ts]: Fixed in 15287a95c. Same-origin embedded frames now resolve one safe-storage adapter from the top window, so they share the adapter’s in-memory fallback when durable storage fails. Added coverage with two distinct frame-local adapters to verify the second frame resolves the first adapter published on the top host.

**Comment by btriapitsyn** [comment]
The feature is wanted — thanks for picking up #3071. Two changes before merge:

1. **Drop the `window.top` storage adapter** (`resolveToolJsonViewStorage`, the `__openchamberToolJsonViewStorage` global, and its test). Same-origin frames already share `localStorage`, so the adapter only changes behavior in the in-memory-fallback edge case, and a Storage object stashed on the top window is machinery nobody will be able to reason about later. Plain `getSafeStorage()` is enough.
2. **Store the preference the way every other UI preference is stored** — as state in the owning store with the existing persistence path — rather than a bespoke localStorage key. That keeps one discoverable home for preferences and gets desktop-settings round-tripping for free. If there's a reason it must stay out of the store, say so in the PR and we can discuss.

The `ToolPart.tsx` wiring itself (read on mount/output change, write on explicit selection) is right and can stay as is.

**Comment by karimodm** [comment]
Addressed in `ab2ae7fb0` and `1d4fb9336`. I removed the `window.top`/bespoke storage path and moved the preference into `useUIStore` with the existing settings synchronization and autosave flow. I also added the shared server sanitizer allowlist and focused coverage so `summary`, `formatted`, and `raw` survive the backend round-trip. The PR body now describes the current store-based design and validation.

**Comment by karimodm** [comment]
Both maintainer requests and the review-bot blocker are addressed on final HEAD `1d4fb9336`.

The bespoke localStorage and `window.top` adapter are gone. `toolJsonViewMode` now lives in `useUIStore` and follows the existing settings persistence path. The web-server sanitizer now accepts the same three values, with a regression test for the real server round-trip.

Focused server settings, UI persistence, and `ToolPart` tests pass, along with UI type-check and lint. I also replaced the stale PR handoff so it describes the store-based design, live cross-card updates, server contract, and current validation results.

**Comment by karimodm** [comment]
The latest automation verdict is PASS and the PR has `review:ready`.

Its handoff warning came from the body snapshot taken before the REST update. The current body no longer mentions the removed adapter or `toolJsonViewPreference.ts`; it documents `useUIStore`, live updates to already-mounted JSON cards, the web-server settings sanitizer, and the final focused tests. No further code or handoff change is needed for that finding.

**Comment by mattv8** [comment]
This overlaps with the settings and persistence work in #3035, so I'll help review and QC it. Give me 24 hours and I'll report back with findings.

**Comment by mattv8** [comment]
Apologies for taking longer than the 24 hrs I promised... I’ve reviewed the PR and no other code changes are necessary.

@btriapitsyn the code is ready to merge after two final steps:

1. Resolve the current merge conflict with `main`.
2. Wait for the required `checks` workflow to pass on the final commit.

Once both are green, the implementation is ready to merge.

**Comment by karimodm** [comment]
Rebased onto current `main` and force-pushed final HEAD `bfb242449`. GitHub now reports the PR mergeable, and the automated review passed with `review:ready`.

The required workflow passed build, type-check, and lint, then stopped at the existing locale-parity failure: current `main` is missing four Turkish Git-discovery keys. #3251 contains exactly those keys and is fully green with `review:ready`. This PR does not touch localization, so I kept that fix in its focused prerequisite PR rather than duplicating it here.

Once #3251 lands, #3072 only needs a rebase onto the updated `main` to get a clean required-check run. The conflict and all code-review findings on this PR are resolved.

**Comment by mattv8** [comment]
Re-reviewed the rebased head `bfb242449`. No further code changes are needed in this PR.

I confirmed the required workflow failed only on Turkish locale-key parity after build, type-check, and lint passed. The four missing Git-discovery keys are outside this PR’s diff, and #3251 contains exactly that focused fix.

@btriapitsyn this is ready to merge once:
- #3251 lands.
- This branch is rebased onto the updated main.
- The required checks workflow passes on that new head.

**Comment by btriapitsyn** [comment]
Thanks @karimodm for following through on the store and persistence changes. I updated the branch from current main, confirmed the JSON-mode round-trip, and ran the focused tests and workspace checks. The required GitHub checks also passed on the integrated head. Merged; this closes #3071.

Thanks @mattv8 for reviewing the implementation and confirming what remained before merge.
